File: minesweeper/java-minesweeper/src/main/java/it/chicio/minesweeper/field/parser/FieldRowContentValuesChainParser.java
package it.chicio.minesweeper.field.parser;

public class FieldRowContentValuesChainParser extends FieldRowContentChainParser {
    boolean canParse(String row) {
        return isNotHeader(row) && isNotTermination(row);
    }

    public FieldsParsingStatus parse(FieldsParsingStatus fieldsParsingStatus) throws RuntimeException {
        FieldsParsingStatus newFieldsParsingStatus = new FieldsParsingStatus(fieldsParsingStatus);
        String[] columnsValues = newFieldsParsingStatus.currentRowContent.split(" ");
        checkIfNumberOfValuesAreCorrect(newFieldsParsingStatus, columnsValues);
        setNewFieldRow(newFieldsParsingStatus, columnsValues);
        return newFieldsParsingStatus;
    }

    private void setNewFieldRow(FieldsParsingStatus newFieldsParsingStatus, String[] columnsValues) {
        for (int column = 0; column < newFieldsParsingStatus.currentField.numberOfColumn(); column++) {
            newFieldsParsingStatus.currentField.set(
                    newFieldsParsingStatus.currentRow,
                    column,
                    columnsValues[column]
            );
        }
        newFieldsParsingStatus.currentRow++;
    }

    private void checkIfNumberOfValuesAreCorrect(FieldsParsingStatus newFieldsParsingStatus, String[] columnsValues) {
        if (columnsValues.length != newFieldsParsingStatus.currentField.numberOfColumn()) {
            throw new RuntimeException();
        }
    }

    private boolean isNotHeader(String row) {
        return !row.matches("(\\d \\d)");
    }

    private boolean isNotTermination(String row) {
        return !row.equals("0 0");
    }
}

File: minesweeper/kotlin-mockk-minesweeper/src/main/kotlin/it/chicio/minesweeper/field/parser/FieldRowContentValuesChainParser.kt
package it.chicio.minesweeper.field.parser

class FieldRowContentValuesChainParser : FieldRowContentChainParser() {
    override fun canParse(row: String?): Boolean = isNotHeader(row) && isNotTermination(row)

    @Throws(RuntimeException::class)
    override fun parse(fieldsParsingStatus: FieldsParsingStatus): FieldsParsingStatus {
        val newFieldsParsingStatus = fieldsParsingStatus.copy()
        newFieldsParsingStatus
                .currentRowContent
                ?.let {
                    val columnsValues = it.split(" ").toTypedArray()
                    checkIfNumberOfValuesAreCorrect(newFieldsParsingStatus, columnsValues)
                    setNewFieldRow(newFieldsParsingStatus, columnsValues)
                }
        return newFieldsParsingStatus
    }

    private fun setNewFieldRow(newFieldsParsingStatus: FieldsParsingStatus, columnsValues: Array<String>) {
        newFieldsParsingStatus.currentField?.let {
            for (column in 0 until it.numberOfColumn) {
                it[newFieldsParsingStatus.currentRow, column] = columnsValues[column]
            }
            newFieldsParsingStatus.currentRow++
        }
    }

    private fun checkIfNumberOfValuesAreCorrect(newFieldsParsingStatus: FieldsParsingStatus, columnsValues: Array<String>) {
        if (columnsValues.size != newFieldsParsingStatus.currentField!!.numberOfColumn) {
            throw RuntimeException()
        }
    }

    private fun isNotHeader(row: String?): Boolean = !(row?.let { it.matches(Regex("(\\d \\d)")) } ?: true)

    private fun isNotTermination(row: String?): Boolean = row != "0 0"
}